groupby.hpp

```diff
@@ -101,7 +102,7 @@ class iter::impl::GroupProducer {
 
     KeyGroupPair* operator->() {
       set_key_group_pair();
-      return current_key_group_pair_.get();
+      return current_key_group_pair_.operator->();
```

Comment:
I suppose it doesn't actually need to be checked in that case: `&*current_key_group_pair_;` should work and seems better to me.
Other than that this all looks fine with the one exception that `DerefHolder` no longer needs user-defined or user-declared constructors, assignment, or destructor.